__tests__/bezier.test.ts

```diff
@@ -0,0 +1,237 @@
+      [1, -1000000, 1, -1000000],
+      [1, -1000000, 1, 0],
+    ])(`Bezier(%d,%d,%d,%d)`, (a, b, c, d) => {
+      expect(Bezier(a, b, c, d)(0)).toBe(0);
+      expect(Bezier(a, b, c, d)(1)).toBe(1);
+    });
+
+    const MONKEY_TRIES = 1000;
+    let allTestPass = true;
+    repeat(MONKEY_TRIES)(() => {
+      const a = Math.random();
+      const b = 3 * Math.random() - 1;
+      const c = Math.random();
+      const d = 3 * Math.random() - 1;
+      const easing = Bezier(a, b, c, d);
+
+      const satisfiesCondition = easing(0) === 0 && easing(1) === 1;
+      if (!satisfiesCondition) {
+        allTestPass = false;
+        test(`Bezier(${a},${b}, ${c}, ${d})(0) = 0 \n Bezier(${a},${b}, ${c}, ${d})(1) = 1`, () => {
+          expect(easing(0)).toBe(0);
+          expect(easing(1)).toBe(1);
+        });
+      }
+    });
+    test(`All ${MONKEY_TRIES} monkey tests pass`, () => {
+      expect(allTestPass).toBeTruthy();
+    });
+  });
+
+  describe('Bezier(a,b,c,d) and Bezier(b,a,d,c) are symmetric about the axis f(x)=x', () => {
+    // The b1 = bezier(a,b,c,d) curve is defined through four points: (0,0), (a,b), (c,d) and (1,1)
+    // The b2 = bezier(b,a,d,c) curve is defined through four points: (0,0), (b,a), (d,c) and (1,1)
+    // These two bezier curves are each others reflection relative to line f(x)=x
+    // So if b1(x) = y, then b2(y) = x and b2(b1(x)) = x
+
+    test.each([
+      [0.0000001, 0, 0.0000001, 0, 0.00000001],
+      [0.0000001, 0, 0.005, 0, 0.00000001],
+      [0.1, 0.2, 0.05, 0.7, 0.00000001],
+      [0.98, 0.45, 0.05, 0.17, 0.00000001],
+      [0.98, 0.45, 0.85, 0.17, 0.00000001],
+
+      // Precision drop when c gets close to one (same drop when d gets close to one due to symmetry)
+      // may be worth to investigate it in the future
+      [0.98, 0.45, 0.9, 0.17, 0.00000001],
+      [0.98, 0.45, 0.99, 0.17, 0.00000001],
+      [0.98, 0.45, 0.999, 0.17, 0.000001],
+      [0.98, 0.45, 0.9999, 0.17, 0.01],
+      [0.98, 0.45, 1, 0.17, 0.01],
+
+      // Precision drop when a gets close to zero (same drop when b gets close to one due to symmetry)
+      [0.99, 0.45, 0.85, 0.17, 0.00000001],
+      [0.99, 0.45, 0.85, 0.17, 0.00000001],
+      [0.999, 0.45, 0.85, 0.17, 0.00000001],
+      [0.9999, 0.45, 0.85, 0.17, 0.00000001],
+      [0.99999, 0.45, 1, 0.85, 0.01],
+      [1, 0.45, 1, 0.85, 0.01],
+
+      [0, 0, 1, 0, 0.01],
+      [1, 1, 0, 1, 0.1],
+      [0, 1, 0, 1, 0.1],
+    ])(`Bezier(%d,%d,%d,%d)`, (a, b, c, d, precision) => {
+      const easing1 = Bezier(a, b, c, d);
+      const easing2 = Bezier(b, a, d, c);
+      expect(
+        functionsAreEqual(
+          (x) => easing1(easing2(x)),
+          (x: number) => x,
+          precision
+        )
+      ).toBeTruthy();
+    });
+
+    const MONKEY_TRIES = 20000;
+    const PRECISION_0 = 0.01;
+    let allTestPass0 = true;
+    repeat(MONKEY_TRIES)(() => {
+      const a = Math.random();
+      const b = Math.random();
+      const c = Math.random();
```

Comment:
Is it something that we agreed that we do in tests? Random may cause tests to be flaky, imo it should be better to have hardcoded few corner cases than random.

Reply:
> Is it something that we agreed that we do in tests?

Yes, @piaskowyk and I were talking about it and decided that it is OK as long as we have hardcoded corner cases too. It is a controversial testing technique, but its not uncommon. These particular tests are verifying bezier function, which has implementation copy-pasted from a linked external library. The tests are based on tests common in that library.

>  Random may cause tests to be flaky, imo it would be better to have hardcoded few corner cases than random.

You may notice, that there is quite a lot of repeat in random tests, each is run 1000 times, this was to reduce the risk of flakiness - when you have 1000 tries it is much easier to catch some error if it occur.

Few corner cases can't be as reliable as 1000 cases. We could hardcode 1000 random values, but this would be much more complicated.

Reply:
>  We could hardcode 1000 random values, but this would be more complicated.
That's definitely something we wouldn't like to do :D The only thing that could be added, or maybe is added (?), is constant seed then random values will be "hardcoded". But if this is something we agreed and 
>as long as we have hardcoded corner cases too 
then it looks ok to me 🎉